### PR TITLE
Handle API errors to trigger force upgrades of the app 

### PR DESCRIPTION
**Pull request by marcaaron**
cc @grgia in case you want to help review?

cc @puneetlath if you can double check the changes where I am removing the migration for the `participants` key 🙏 

### Details

Adds an update required view and allows for deprecating older App versions.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/340997

### Tests

**Setup:**

Apply this diff so that the upgrade works. It will be disabled on dev and staging:

```diff
diff --git a/src/libs/actions/AppUpdate/index.ts b/src/libs/actions/AppUpdate/index.ts
index 7db334919b..d76b89f6ed 100644
--- a/src/libs/actions/AppUpdate/index.ts
+++ b/src/libs/actions/AppUpdate/index.ts
@@ -17,9 +17,9 @@ function triggerUpgradeRequired() {
     // We should only update the minimum app version in the API after all platforms of a new version have been deployed to PRODUCTION.
     // As staging is always ahead of production there is no reason to "force update" those apps.
     getEnvironment().then((environment) => {
-        if (environment !== CONST.ENVIRONMENT.PRODUCTION) {
-            return;
-        }
+        // if (environment !== CONST.ENVIRONMENT.PRODUCTION) {
+        //     return;
+        // }

         Onyx.set(ONYXKEYS.UPDATE_REQUIRED, true);
     });
```

1. In Web-Expensify modify the `Request::MINIMUM_APP_VERSION` variable here so that it is _higher_ than the current app version.
2. Navigate to NewDot (fully logged out)
3. Try to log in
4. Verify you get the "Update required" view
5. Reset the `Request::MINIMUM_APP_VERSION`
6. Refresh app/page
7. Verify you do not see the "Update required" view anymore
8. Sign in
9. Modify the `Request::MINIMUM_APP_VERSION` variable here so that it is _higher_ than the current app version
10. Take some action in the app
11. Verify you get the "Update required" view

**Test "Update" button behavior**

Web/mWeb - App should refresh page
Android - App should go to the Play Store
iOS - App should go to the App Store
Desktop - App should auto 

- [ ] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

QA is not possible for these changes.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos

Note - I tested on all platforms. However, there were some changes to the body copy so the screenshots do not 100% reflect where we landed.

<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by marcaaron** [comment]
Will do some testing for this one tomorrow and get some design feedback.

File: src/libs/migrations/PersonalDetailsByAccountID.js
Comment (marcaaron) [inline src/libs/migrations/PersonalDetailsByAccountID.js:258]: This is getting removed as we will be using it again soon and do not want the migration to unset the value. I am pretty sure this should be fine for now as nothing should be using or sending the old `participants` array.

**Comment by marcaaron** [comment]
Working well on most platforms except Android which is completely untestable locally. Not sure why yet.

**Comment by marcaaron** [comment]
There is some conversation still happening around what the best body copy should be for this. But gonna pop it into review since I _think_ it's almost there.

**Comment by marcaaron** [comment]
@c3024 it won't be possible to test this on native without code changes. For web you can use network request overrides in ChromeDevTools to force the JSON code. For other platforms you can trigger the error manually by throwing in HttpUtils or elsewhere or setting the Onyx data manually. LMK if you need any assistance though.

File: src/libs/actions/UpdateRequired.ts
Comment (grgia) [inline src/libs/actions/UpdateRequired.ts:10]: Any specific reasons to default to DEV here?

**Review by grgia — COMMENTED** [review]
image

is there a reason IOS native screenshot doesnt have the upper left hand corner "update required" text?

File: src/styles/index.ts
Comment (grgia) [inline src/styles/index.ts:4171]: shall we add these two to `variables` too?

File: src/styles/utils/index.ts
Comment (grgia) [inline src/styles/utils/index.ts:1438]: same with this 160, maybe add to variables?

File: tests/unit/MigrationTest.js
Comment (grgia) [inline tests/unit/MigrationTest.js:432]: Why are we removing this block?

File: src/libs/migrations/PersonalDetailsByAccountID.js
Comment (puneetlath) [reply src/libs/migrations/PersonalDetailsByAccountID.js:258]: Ok cool that seems fine to me.

File: src/libs/actions/UpdateRequired.ts
Comment (marcaaron) [reply src/libs/actions/UpdateRequired.ts:10]: Not really! I think I copied this from somewhere. It could be an empty string. Just want to make suer we are on production and not confusing people on staging or dev.

File: tests/unit/MigrationTest.js
Comment (marcaaron) [reply tests/unit/MigrationTest.js:432]: Oh! Right, it's the same reason as provided here: https://github.com/Expensify/App/pull/32326/files#r1445438321

**Comment by marcaaron** [comment]
Fixed conflicts. @c3024 please review when you get a chance 🙇

**Comment by c3024** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by marcaaron** [comment]
@c3024 are you still helping out with the testing here?

**Comment by c3024** [comment]
Yes, I am not able to test it on mWeb. Tests fine on desktop and web. I'm trying again now.

**Comment by marcaaron** [comment]
>I am not able to test it on mWeb

Can you tell me why? I think you can force the screen to show and confirm the styles look OK? Let me know how I can help.

**Comment by c3024** [comment]
I got the screen with overriding network request in Chrome Devtools for Web Chrome and directly throwing the error in HttpUtils for Desktop. 

I tried throwing the error directly in HttpUtils with iOS Safari but the update screen does not show up. I think the error is caught somewhere up? Let me try again once. I'll get back if I need help. I will finish this up today. Thanks.

**Comment by c3024** [comment]
Space between the 'Update required' text and the animation on the top and the 'Update' button and the bottom of the screen is asymmetric.

Android mWeb

Screenshot_2024-01-19-10-34-03-948_com android chrome

iOS mWeb

updateRequirediOSSafari

iOS Native

Screenshot 2024-01-19 at 10 39 34 AM

**Review by c3024 — APPROVED** [review]
LGTM

**Comment by marcaaron** [comment]
Thanks for the review! @c3024 

>Space between the 'Update required' text and the animation on the top and the 'Update' button and the bottom of the screen is asymmetric.

This is the same in the mock provided by design so I _believe_ the intention is for the main view to be centered.

I think it looks pretty decent - people will see this very very very infrequently in any case 😄 

But speak now or forever hold your peace

**Comment by marcaaron** [comment]
This one's just waiting for a final review from an internal reviewer if @puneetlath @grgia @MonilBhavsar have a chance to give a final review. 🙇

File: src/libs/migrations/PersonalDetailsByAccountID.js
Comment (MonilBhavsar) [reply src/libs/migrations/PersonalDetailsByAccountID.js:258]: > This is getting removed as we will be using it again soon and do not want the migration to unset the value.

Sorry, I didn't get. What we will be using again soon?

> I am pretty sure this should be fine for now as nothing should be using or sending the old participants array.

This looks good 👍

**Comment by marcaaron** [comment]
Bump for reviews.

**Review by MonilBhavsar — COMMENTED** [review]
@marcaaron I commented here yesterday https://github.com/Expensify/App/pull/32326/files#r1461660832
That's the only clarification I need. Looks good overall

File: src/libs/migrations/PersonalDetailsByAccountID.js
Comment (marcaaron) [reply src/libs/migrations/PersonalDetailsByAccountID.js:258]: the `report.participants` value - but with a different data structure.

**Comment by marcaaron** [comment]
@puneetlath @grgia would one of you mind reviewing and/or approving this? Checked out @MonilBhavsar's question and don't think it's a blocker.

**Review by puneetlath — APPROVED** [review]
Looks good to me!

**Comment by hannojg** [comment]
This PR broke my development environment, as it introduced a bunch (~20) of require cycles (thus making modules undefined).

```
 WARN  Require cycle: src/libs/Log.ts -> src/libs/Network/index.ts -> src/libs/Network/MainQueue.ts -> src/libs/Request.ts -> src/libs/HttpUtils.ts -> src/libs/actions/UpdateRequired.ts -> src/libs/Environment/getEnvironment/index.native.ts -> src/libs/Environment/betaChecker/index.android.ts -> src/libs/actions/AppUpdate/index.ts -> src/libs/actions/AppUpdate/updateApp/index.android.ts -> src/libs/actions/Link.ts -> src/libs/API.ts -> src/libs/Log.ts

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: src/libs/Log.ts -> src/libs/Network/index.ts -> src/libs/Network/MainQueue.ts -> src/libs/Request.ts -> src/libs/HttpUtils.ts -> src/libs/actions/UpdateRequired.ts -> src/libs/Environment/getEnvironment/index.native.ts -> src/libs/Environment/betaChecker/index.android.ts -> src/libs/actions/AppUpdate/index.ts -> src/libs/actions/AppUpdate/updateApp/index.android.ts -> src/libs/actions/Link.ts -> src/libs/API.ts -> src/libs/Middleware/index.ts -> src/libs/Middleware/Logging.ts -> src/libs/Log.ts
```
Screenshot_1706085299

First documented here: 
https://expensify.slack.com/archives/C01GTK53T8Q/p1706084435877639

**Comment by mountiny** [comment]
@marcaaron I went ahead with straight revert to unblock the testing environment nad fix the crash.

Seems like there must have been some issue after merging with main as you have tested on android and it clearly worked. Hopefully it will be quick fix

**Comment by c3024** [comment]
There are two cycles.

1. Using getEnvironment in UpdateRequired.ts: We might fix this with moving [this check](https://github.com/Expensify/App/blob/c509c20cc3334862545363dd8272e1209440b823/src/libs/actions/UpdateRequired.ts#L10-L13) in UpdateRequired.ts to [here](https://github.com/Expensify/App/blob/c509c20cc3334862545363dd8272e1209440b823/src/Expensify.js#L213) in Expensify.js
2. Using Link in updateApp/index.android.ts: We might fix this by using Linking.openURL using Linking from react-native instead of Link.openExternalLink from userActions/Link

**Comment by marcaaron** [comment]
Oh great 😄

**Comment by marcaaron** [comment]
@c3024 Did you test on Android here? Curious how we missed this one..

**Comment by marcaaron** [comment]
I remember fixing these issues at one point. And they came back. I think maybe when I tried to resolve some unverified commits I must have reintroduced the bad code.

**Pull request merged**
